### PR TITLE
Add automated story-based static accessibility testing workflow and add in-console test documentation

### LINKED ISSUE #1180: Add automated story-based static accessibility testing framework
We have initial reviews and we have calendar reviews every so often. We may want to look into additional a11y tests in our CI. https://github.com/dequelabs/axe-core

Since the aXe test is of static DOM/HTML, we should probably try to run it in Jest on all the https://github.com/dequelabs/axe-core/tree/develop/doc/examples/jest_react 

Mocha is a possibility though. It does look like aXe requires JSDOM, so we may need to re-enable that in Jest.

This looks promising for Jest/aXe testing
https://github.com/nickcolley/jest-axe#readme

Looks like we can enable JSDOM for just the accessibility tests. https://jestjs.io/docs/en/configuration.html#testenvironment-string

### PR DESCRIPTION
**Pull request by interactivellama**
Fixes #1180.

### Additional description
This does not add the testing to the CI, just the ability to run testing with `npm run test:a11y`. Adding the aXe audit will more than double the TravisCI time.

* Two Jest configs are needed since many components do not pass in Jest/jsDOM snapshot testing, but due in Jest/Node environment (such as portal use). aXe needs a DOM and therefore uses jsDOM
* Add `npm run test:accessibility` script
* Moves default Jest config to preset file
* Removes Webpack build, doc build line separator, and passing Mocha tests from the CI console--making it more concise.
* Removes IconSettings tests and Tree's Large dataset (300+) from Jest tests
* Guards SVGs, SLDSSettings.setAppElement and canvas functions for use in jsDOM environment 

---

### CONTRIBUTOR checklist (do not remove)
Please complete for every pull request

* [ ] First-time contributors should sign the Contributor License Agreement. It's a fancy way of saying that you are giving away your contribution to this project. If you haven't before, wait a few minutes and a bot will comment on this pull request with instructions.
* [ ] `npm run lint:fix` has been run and linting passes.
* [ ] Mocha, Jest (Storyshots), and `components/component-docs.json` CI tests pass (`npm test`).
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).

### REVIEWER checklist (do not remove)

* [ ] TravisCI tests pass. This includes linting, Mocha, Jest, Storyshots, and `components/component-docs.json` tests.
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).
###### Required only if there are markup / UX changes
* [ ] Add year-first date and commit SHA to `last-slds-markup-review` in `package.json` and push.
* [ ] Request a review of the deployed Heroku app by the Salesforce UX Accessibility Team.
* [ ] Add year-first review date, and commit SHA, `last-accessibility-review`, to `package.json` and push.
* [ ] While the contributor's branch is checked out, run `npm run local-update` within locally cloned [site repo](https://github.com/salesforce-ux/design-system-react-site) to confirm the site will function correctly at the next release.

### CODE REVIEW

**Comment by futuremint** [comment]
Running this locally shows 72 errors.  Are we going to fix those are part of this PR or fix those in a separate PR?

**Comment by interactivellama** [comment]
I propose a separate PR(s) that is grouped with the findings from the upcoming audit from JoAnne

**Comment by interactivellama** [comment]
The other thing we need to talk about is when this will run since it takes so long.

File: package.json [hunk 1]
```diff
@@ -35,7 +35,8 @@
 		"test:docs": "npx babel-node ./tests/component-docs-test.js",
 		"test:fix": "npm run lint:fix && npm run test",
 		"test:unit": "cross-env NODE_ENV=test npx karma start --single-run",
-		"test:snapshot": "cross-env NODE_ENV=test npm run storyshots:build && npx jest",
+		"test:snapshot": "cross-env NODE_ENV=test npm run storyshots:build && npx jest --config=tests/snapshots-config.json",
+		"test:a11y": "cross-env NODE_ENV=test npx jest --config=tests/a11y-config.json",
```
[the remark this answers is not in the document]
Comment (interactivellama) [reply package.json]: I've made `test:a11y` an alias

File: components/story-based-tests.js [hunk 2]
```diff
@@ -2,13 +2,13 @@
...
- * snapshots. With enough Storybook story examples, these two methods should be
- * significant to fully test components with the exception of callback testing.
- * Pleaes test callback props with the Mocha framework (http://localhost:8001).
- * `tests/story-based-tests.snapshot-test.js` looks at this file and uses the
- * following stories to create snapshots.
+ * as the basis for DOM and image snapshots to allow markup and visual
+ *  regression testing. All new components should be added below to enable 
+ * DOM and image snapshots. With enough Storybook story examples, these two
+ * methods should be significant to fully test components with the exception
+ * of callback testing. Pleaes test callback props with the Mocha framework
```
[the remark this answers is not in the document]
Comment (interactivellama) [reply components/story-based-tests.js]: Fixed

File: tests/story-based-accessibility-tests.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ * For more information, please visit:
+ * https://github.com/storybooks/storybook/tree/master/addons/storyshots
+ */
+
+// Ran in Jest jsDOM environment
+
+import { axe, toHaveNoViolations } from 'jest-axe';
+import { configure, mount } from 'enzyme';
+import Adapter from 'enzyme-adapter-react-16';
+import initStoryshots, { imageSnapshot } from '@storybook/addon-storyshots';
```
[the remark this answers is not in the document]
Comment (interactivellama) [reply tests/story-based-accessibility-tests.js]: Removed

File: tests/story-based-tests.snapshot-test.js [hunk 4]
```diff
@@ -30,6 +31,7 @@ app.use(express.static(`${rootPath}/storybook-based-tests`));
 // Create DOM snapshot tests from Storybook stories
 initStoryshots({
 	configPath: '.storybook-based-tests',
+	storyNameRegex: /^((?!.*?(NoTest)).)*$/,
```
[the remark this answers is not in the document]
Comment (interactivellama) [reply tests/story-based-tests.snapshot-test.js]: Added.

File: tests/story-based-tests.snapshot-test.js [hunk 5]
```diff
@@ -66,6 +68,7 @@ describe('Image Snapshots', function imageSnapshotFunction() {
 	// snapshot tests.
 	initStoryshots({
 		configPath: '.storybook-based-tests',
+		storyNameRegex: /^((?!.*?(NoTest)).)*$/,
```
[the remark this answers is not in the document]
Comment (interactivellama) [reply tests/story-based-tests.snapshot-test.js]: Done.

File: utilities/warning/check-app-element-set.js [hunk 6]
```diff
@@ -12,7 +13,7 @@ if (process.env.NODE_ENV !== 'production') {
 		const appElement = Settings.getAppElement();
 		/* eslint-disable max-len */
 		warning(
-			!!appElement,
+			canUseDOM && !!appElement,
```
[the remark this answers is not in the document]
Comment (interactivellama) [reply utilities/warning/check-app-element-set.js]: Updated.

**Comment by interactivellama** [comment]
screen shot 2019-02-16 at 4 24 55 pm

@futuremint Really good requests! Now, that we've got the 404 and console error noise down, in headers with context and additional documentation is a really good features to have for folks running into errors.

**Comment by interactivellama** [comment]
Issue added to fix errors https://github.com/salesforce/design-system-react/issues/1732

**Pull request merged**